ompi/mca/coll/xhc/coll_xhc.c

```diff
@@ -0,0 +1,700 @@
+/*
+ * Copyright (c) 2021-2023 Computer Architecture and VLSI Systems (CARV)
+ *                         Laboratory, ICS Forth. All rights reserved.
+ * $COPYRIGHT$
+ *
+ * Additional copyrights may follow
+ *
+ * $HEADER$
+ */
+
+#include "ompi_config.h"
+
+#include "mpi.h"
+#include "ompi/constants.h"
+#include "ompi/communicator/communicator.h"
+#include "ompi/mca/coll/coll.h"
+#include "ompi/mca/coll/base/base.h"
+
+#include "opal/mca/rcache/rcache.h"
+#include "opal/mca/shmem/base/base.h"
+#include "opal/mca/smsc/smsc.h"
+
+#include "opal/include/opal/align.h"
+#include "opal/util/show_help.h"
+#include "opal/util/minmax.h"
+
+#include "coll_xhc.h"
+
+static int xhc_make_comms(ompi_communicator_t *ompi_comm,
+    xhc_peer_info_t *peer_info, xhc_comm_t **comms_dst,
+    int *comm_count_dst, xhc_loc_t *hierarchy, int hierarchy_len);
+static void xhc_destroy_comms(xhc_comm_t *comms, int comm_count);
+
+static void xhc_print_info(xhc_module_t *module,
+    ompi_communicator_t *comm, xhc_data_t *data);
+
+static void *xhc_shmem_create(opal_shmem_ds_t *seg_ds, size_t size,
+    ompi_communicator_t *ompi_comm, const char *name_chr_s, int name_chr_i);
+static void *xhc_shmem_attach(opal_shmem_ds_t *seg_ds);
+static mca_smsc_endpoint_t *xhc_smsc_ep(xhc_peer_info_t *peer_info);
+
+int xhc_lazy_init(xhc_module_t *module, ompi_communicator_t *comm) {
+
+    int comm_size = ompi_comm_size(comm);
+    int rank = ompi_comm_rank(comm);
+
+    xhc_peer_info_t *peer_info = module->peer_info;
+
+    opal_shmem_ds_t *peer_cico_ds = NULL;
+    xhc_data_t *data = NULL;
+
+    int return_code = OMPI_SUCCESS;
+    int ret;
+
+    errno = 0;
```

Comment:
What is the question here? Why errno is set to 0? If so, it's for error reporting: in case something fails, I show what errno indicated. In the past I've had trouble where errno was set to non-zero outside the code in question, and when the code failed in a way that errno is not altered, the stale value of errno mileadingly hinted to another error. I don't have strong feelings about it -- let me know if it would be best to remove it.

Reply:
we report vie the return code not via errno. Moreover,  why setting it to success ?

Reply:
I also report init errors via a help text -- perhaps that could/should go away. In the olden times before smsc, xhc_init also handled xpmem initialization, so failures there were even more common due to xpmem misconfigurations, and it was helpful to be explicitly notified of them. So the help text there is in part a remnant that could potentially be removed.

Or maybe it should only appear when e.g. the priority is explicitly set to non-zero. Or maybe something like that would be nice in coll/base -- a mechanism to inform the user that something went wrong with something that they explicitly requested. Anyway.

The errno set to 0 is to bring it to a known state. Suppose that some previously running code encounters an error, e.g. a failed open for a non-existing file, but does not abort because of it. My init may fail because of some unrelated reason, but errno will misleadingly indicate that failure is related to `No such file or directory`.